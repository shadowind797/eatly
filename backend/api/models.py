from django.db import models
from django.contrib.auth.models import AbstractUser


class User(AbstractUser):
    username = models.CharField(max_length=120, unique=True)
    password = models.CharField(max_length=120)
    phone = models.CharField(max_length=120, blank=True)
    status = models.ForeignKey('Status', on_delete=models.PROTECT, default=5)
    is_banned = models.BooleanField(default=False)
    ban_reason = models.TextField(blank=True)
    passwd_change_link = models.CharField(max_length=120, blank=True)

    def __str__(self):
        return self.username


class Item(models.Model):
    title = models.CharField(max_length=120, unique=True)
    description = models.TextField()
    price = models.FloatField()
    photo = models.ImageField(upload_to='img')
    category = models.ForeignKey('Category', on_delete=models.SET_NULL, null=True, blank=True)
    restaurant = models.ForeignKey('Restaurant', on_delete=models.SET_NULL, null=True, blank=True)
    rating = models.FloatField()

    def __str__(self):
        return self.title


class Restaurant(models.Model):
    name = models.CharField(max_length=120, unique=True)
    address = models.CharField(max_length=120)
    rating = models.FloatField()
    image = models.ImageField(upload_to='img/rests', null=True)
    category = models.ForeignKey('RestaurantCat', on_delete=models.SET_NULL, null=True, blank=True)

    def __str__(self):
        return self.name


class Address(models.Model):
    house_address = models.CharField(max_length=120)
    entrance = models.CharField(max_length=120)
    floor = models.CharField(max_length=120)
    flat = models.CharField(max_length=120)
    owner = models.ForeignKey('User', on_delete=models.CASCADE)
    active = models.BooleanField(default=False)

    def __str__(self):
        return self.house_address


class Order(models.Model):
    user = models.ForeignKey('User', on_delete=models.SET_NULL, null=True)
    payment = models.ForeignKey('Payments', on_delete=models.SET_NULL, null=True)
    address = models.ForeignKey('Address', on_delete=models.SET_NULL, null=True)
    total = models.FloatField()
    status = models.ForeignKey('OrderStatus', on_delete=models.SET_NULL, null=True)
    restaurant = models.ForeignKey('Restaurant', on_delete=models.SET_NULL, null=True)
    coupon = models.ForeignKey('Coupon', on_delete=models.SET_NULL, null=True)
    comment = models.TextField(null=True, blank=True)

    def __str__(self):
        return self.user


class OrderStatus(models.Model):
    name = models.CharField(max_length=120, unique=True)


class Coupon(models.Model):
    title = models.CharField(max_length=120, unique=True)
    valid_from = models.DateField(auto_now_add=True)
    valid_to = models.DateField()
    is_valid = models.BooleanField(default=True)
    value = models.FloatField()
    times_activated = models.IntegerField(default=0)
    enabled_times = models.IntegerField()
    category = models.ForeignKey('CouponCategory', on_delete=models.CASCADE)

    def __str__(self):
        return self.title


class CouponCategory(models.Model):
    name = models.CharField(max_length=120, unique=True)

    def __str__(self):
        return self.name


class Payments(models.Model):
    number = models.CharField(max_length=120, unique=True)
    date_to = models.CharField(max_length=100)
    cvv = models.IntegerField()
    owner = models.ForeignKey('User', on_delete=models.CASCADE)
    active = models.BooleanField(default=False)
    name = models.CharField(max_length=120, default="")

    def __str__(self):
        return self.number


class CartItem(models.Model):
    item = models.ForeignKey('Item', on_delete=models.CASCADE)
    quantity = models.IntegerField(default=1)
    owner = models.ForeignKey('User', on_delete=models.CASCADE)

    def __str__(self):
        return self.item.title


class Reviews(models.Model):
    author = models.ForeignKey('User', on_delete=models.CASCADE)
    time = models.CharField(max_length=120)
    text = models.TextField()
    rating = models.FloatField()
    created_at = models.DateTimeField(auto_now_add=True)

    def __str__(self):
        return self.author.username


class RestaurantCat(models.Model):
    name = models.CharField(max_length=120, unique=True)

    def __str__(self):
        return self.name


class Category(models.Model):
    name = models.CharField(max_length=120, unique=True)

    def __str__(self):
        return self.name


class Status(models.Model):
    name = models.CharField(max_length=120)

    def __str__(self):
        return self.name
