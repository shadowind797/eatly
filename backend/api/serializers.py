from .models import *
from rest_framework import serializers


class UserSerializer(serializers.ModelSerializer):
    class Meta:
        model = User
        fields = ['id', 'username', 'email', 'password', "phone", "status", "is_banned", "ban_reason", "first_name"]
        extra_kwargs = {"password": {"write_only": True}, "status": {"read_only": True}}

    def create(self, validated_data):
        user = User.objects.create_user(**validated_data)
        return user


class AddressSerializer(serializers.ModelSerializer):
    class Meta:
        model = Address
        fields = ["id", "house_address", "entrance", "floor", "flat"]

    def create(self, validated_data):
        address = Address.objects.create_address(**validated_data)
        return address


class PaymentSerializer(serializers.ModelSerializer):
    class Meta:
        model = Payments
        fields = "__all__"
        extra_kwargs = {"cvv": {"write_only": True}, "date_to": {"read_only": True}}

    def create(self, validated_data):
        payment = Payments.objects.create_payments(**validated_data)
        return payment


class OrderSerializer(serializers.ModelSerializer):
    class Meta:
        model = Order
        fields = ["id", "status", "total", "payment", "address"]

    def create(self, validated_data):
        order = Order.objects.create_order(**validated_data)
        return order


class ItemSerializer(serializers.ModelSerializer):
    class Meta:
        model = Item
        fields = ['id', "title", "description", "price", "photo", "category", "rating"]


class CartItemSerializer(serializers.ModelSerializer):
    class Meta:
        model = CartItem
        fields = ['id', "item", "quantity"]


class CategorySerializer(serializers.ModelSerializer):
    class Meta:
        model = Category
        fields = ['id', "name"]


class RestaurantSerializer(serializers.ModelSerializer):
    class Meta:
        model = Restaurant
        fields = ['id', "name", "address", "rating", "category_id", "image"]


class RestaurantCatSerializer(serializers.ModelSerializer):
    class Meta:
        model = RestaurantCat
        fields = ['id', "name"]
