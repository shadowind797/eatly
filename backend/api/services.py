from django.conf import settings
from django.shortcuts import redirect
from django.core.exceptions import ValidationError
from urllib.parse import urlencode
from typing import Dict, Any
import requests
from .models import User
import jwt

GOOGLE_ACCESS_TOKEN_OBTAIN_URL = 'https://oauth2.googleapis.com/token'
GOOGLE_USER_INFO_URL = 'https://www.googleapis.com/oauth2/v3/userinfo'

GITHUB_ACCESS_TOKEN_OBTAIN_URL = 'https://github.com/login/oauth/access_token'
GITHUB_USER_INFO_URL = 'https://api.github.com/user'

LOGIN_URL = f'{settings.BASE_APP_URL}/login'


def google_get_access_token(code: str, redirect_uri: str) -> str:
    data = {
        'code': code,
        'client_id': settings.GOOGLE_OAUTH2_CLIENT_ID,
        'client_secret': settings.GOOGLE_OAUTH2_CLIENT_SECRET,
        'redirect_uri': redirect_uri,
        'grant_type': 'authorization_code'
    }

    response = requests.post(GOOGLE_ACCESS_TOKEN_OBTAIN_URL, data=data)
    if not response.ok:
        raise ValidationError('Could not get access token from Google.')

    access_token = response.json()['access_token']

    return access_token


def google_get_user_info(access_token: str) -> Dict[str, Any]:
    response = requests.get(
        GOOGLE_USER_INFO_URL,
        params={'access_token': access_token}
    )

    if not response.ok:
        raise ValidationError('Could not get user info from Google.')

    return response.json()


def github_get_access_token(code: str, redirect_uri: str) -> str:
    headers = {
        "Accept": "application/json"
    }
    data = {
        'code': code,
        'client_id': settings.GITHUB_OAUTH2_CLIENT_ID,
        'client_secret': settings.GITHUB_OAUTH2_CLIENT_SECRET,
        'redirect_uri': redirect_uri
    }

    response = requests.post(GITHUB_ACCESS_TOKEN_OBTAIN_URL, headers=headers, data=data)
    if not response.ok:
        raise ValidationError('Could not get access token from Github.')

    access_token = response.json()['access_token']

    return access_token


def github_get_user_info(access_token: str) -> Dict[str, Any]:
    headers = {
        'Authorization': f'Bearer {access_token}'
    }
    response = requests.get(
        GITHUB_USER_INFO_URL,
        headers=headers
    )

    if not response.ok:
        raise ValidationError('Could not get user info from Github.')

    return response.json()


def get_user_data(validated_data, platform):
    domain = settings.BASE_API_URL
    redirect_uri = f'{domain}/api/login/{platform}/'

    code = validated_data.get('code')
    error = validated_data.get('error')
    state = validated_data.get('state')

    if error or not code:
        params = urlencode({'error': error})
        return redirect(f'{LOGIN_URL}?{params}')

    if platform == "google":
        access_token = google_get_access_token(code=code, redirect_uri=redirect_uri)
        user_data = google_get_user_info(access_token=access_token)

        user, created = User.objects.get_or_create(
            email=user_data['email'],
        )

        if not created:
            user.username = user_data['name']
            user.save()

        profile_data = {
            'email': user_data['email'],
        }
        return profile_data
    elif platform == "github":
        access_token = github_get_access_token(code=code, redirect_uri=redirect_uri)
        user_data = github_get_user_info(access_token=access_token)

        print(user_data)

        User.objects.get_or_create(
            username=user_data['login'],
        )

        profile_data = {
            'username': user_data['login'],
        }
        return profile_data
