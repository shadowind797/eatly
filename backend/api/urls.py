from django.urls import path, include

from . import views

urlpatterns = [
    path('access/', views.GetAccess.as_view(), name='access'),
    path('user/', views.GetUser.as_view(), name='user'),
    path('user/change/', views.GetUser.as_view(), name='user-change'),

    path("login/google/", views.GoogleLoginApi.as_view(), name='google-login'),
    path("login/github/", views.GithubLoginApi.as_view(), name='github-login'),

    path('items/', views.ItemListCreate.as_view(), name='item-list'),
    path('items/top/', views.ItemListCreate.as_view(), name='item-list'),
    path('items/category/', views.CategoriesList.as_view(), name='item-category-list'),

    path('items/cart/', views.CartItemListCreate.as_view(), name='item-cart-list'),
    path('items/cart/add', views.CartItemListCreate.as_view(), name='item-cart-list'),
    path('items/cart/delete', views.DeleteCartItem.as_view(), name='item-cart-list'),
    path('items/cart/check', views.CheckInCart.as_view(), name='item-cart-list'),

    path("items/delete/<int:pk>/", views.ItemDelete.as_view(), name='item-delete'),

    path("restaurants/", views.RestaurantListCreate.as_view(), name='restaurant-list'),
    path("restaurants/categories/", views.RestaurantCategories.as_view(), name='restaurant-cats-list'),

    path("address/", views.AddressList.as_view(), name='address-list'),
    path("address/add/", views.AddressList.as_view(), name='restaurant-cats-list'),

    path("order/", views.OrderView.as_view(), name='order'),
    path("order/add/", views.OrderView.as_view(), name='order-create'),
    path("order/cancel/", views.CancelOrder.as_view(), name='order-cancel'),

    path("payment/", views.PaymentView.as_view(), name='payment'),
    path("payment/add/", views.PaymentView.as_view(), name='payment-create'),

    path("coupon/", views.CouponListCreate.as_view(), name='coupon-list'),
    path("coupon/add/", views.CouponListCreate.as_view(), name='coupon-create'),

    path("items/search/", views.SearchView.as_view(), name='search'),
    path("items/search/filters/", views.FilterView.as_view(), name='filter'),

    path("profile/orders", views.GetProfile.as_view(), name='all-user-orders'),
]
