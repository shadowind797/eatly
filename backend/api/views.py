from django.db.models import Q
from django.utils import timezone
from datetime import datetime
from rest_framework.response import Response

from .models import *
from rest_framework import generics, status
from .serializers import *
from rest_framework.permissions import IsAuthenticated, AllowAny

class GetAccess(generics.ListAPIView):
    serializer_class = UserSerializer
    permission_classes = (IsAuthenticated,)

    def list(self, request, *args, **kwargs):
        access_to = self.request.query_params.getlist("access_to")
        user = self.request.user

        if access_to:
            if "site" in access_to:
                is_banned = User.objects.filter(pk=user.id, is_banned=True).exists()
                if is_banned:
                    return Response(status=status.HTTP_403_FORBIDDEN)
                else:
                    return Response(status=status.HTTP_200_OK)
            elif "admin" in access_to:
                search = User.objects.filter(pk=user.id, status=2).exists()
                search2 = User.objects.filter(pk=user.id, status=1).exists()
                if search or search2:
                    return Response(status=status.HTTP_200_OK)
                else:
                    return Response(status=status.HTTP_403_FORBIDDEN)
            elif "manage" in access_to:
                search = User.objects.filter(pk=user.id, status=3).exists()
                search2 = User.objects.filter(pk=user.id, status=1).exists()
                if search or search2:
                    return Response(status=status.HTTP_200_OK)
                else:
                    return Response(status=status.HTTP_403_FORBIDDEN)
            elif "orders" in access_to:
                search = User.objects.filter(pk=user.id, status=4).exists()
                search2 = User.objects.filter(pk=user.id, status=1).exists()
                if search or search2:
                    return Response(status=status.HTTP_200_OK)
                else:
                    return Response(status=status.HTTP_403_FORBIDDEN)
            elif "header_options" in access_to:
                if user.status.id == 1:
                    return Response(data=[
                        {
                            "pageName": "admin",
                            "slug": "/admin/",
                            "name": "Admin"
                        },
                        {
                            "pageName": "manage",
                            "slug": "/manage/",
                            "name": "Manage"
                        },
                        {
                            "pageName": "orders",
                            "slug": "/orders/",
                            "name": "Orders"
                        }
                    ], status=status.HTTP_200_OK)
                elif user.status.id == 2:
                    return Response(data=[
                        {
                            "pageName": "admin",
                            "slug": "/admin/",
                            "name": "Admin"
                        }
                    ], status=status.HTTP_200_OK)
                elif user.status.id == 3:
                    return Response(data=[
                        {
                            "pageName": "manage",
                            "slug": "/manage/",
                            "name": "Manage"
                        },
                    ], status=status.HTTP_200_OK)
                elif user.status.id == 4:
                    return Response(data=[
                        {
                            "pageName": "orders",
                            "slug": "/orders/",
                            "name": "Orders"
                        }
                    ], status=status.HTTP_200_OK)
                elif user.status.id == 5 or user.status.id == 6:
                    return Response(data=[], status=status.HTTP_200_OK)
                else:
                    return Response(status=status.HTTP_401_UNAUTHORIZED)
        return Response(status=status.HTTP_400_BAD_REQUEST)


class GetUser(generics.ListAPIView):
    serializer_class = UserSerializer
    permission_classes = (IsAuthenticated,)

    def post(self, request, *args, **kwargs):
        method = self.request.data.get("method")
        name = self.request.data.get("name")

        if method == "name":
            user = self.request.user
            userProfile = User.objects.get(pk=user.id)
            userProfile.first_name = name
            userProfile.save()
            return Response(status=status.HTTP_201_CREATED)

    def get_queryset(self):
        user = self.request.user.id
        return User.objects.filter(pk=user)


class PaymentView(generics.ListAPIView):
    serializer_class = PaymentSerializer
    permission_classes = (IsAuthenticated,)

    def get_queryset(self):
        payment_id = self.request.query_params.get("payment_id")
        user = self.request.user

        if payment_id:
            return Payments.objects.filter(pk=payment_id, owner=user)
        else:
            return Payments.objects.filter(owner=user)

    def post(self, request, *args, **kwargs):
        number = self.request.data.get("number")
        date_to = self.request.data.get("date_to")
        cvv = self.request.data.get("cvv")
        name = self.request.data.get("name")

        payment = Payments.objects.filter(owner=self.request.user, number=number, date_to=date_to, cvv=cvv, name=name).exists()

        if payment:
            return Response(status=status.HTTP_409_CONFLICT)
        else:
            new_payment = Payments(owner=self.request.user, number=number, date_to=date_to, cvv=cvv, name=name)
            new_payment.save()
            return Response(status=status.HTTP_201_CREATED)


class AddressList(generics.ListAPIView):
    serializer_class = AddressSerializer
    permission_classes = (IsAuthenticated,)

    def get_queryset(self):
        address_id = self.request.query_params.get("address_id")
        user = self.request.user

        if address_id:
            return Address.objects.filter(pk=address_id, owner=user)
        else:
            return Address.objects.filter(owner=user)

    def post(self, request, *args, **kwargs):
        building_address = self.request.data.get("house_address")
        entrance = self.request.data.get("entrance")
        floor = self.request.data.get("floor")
        flat = self.request.data.get("flat")

        address = Address.objects.filter(owner=self.request.user, house_address=building_address, entrance=entrance, floor=floor, flat=flat).exists()

        if address:
            return Response(status=status.HTTP_409_CONFLICT)
        else:
            new_address = Address(owner=self.request.user, house_address=building_address, entrance=entrance, floor=floor, flat=flat)
            new_address.save()
            return Response(status=status.HTTP_201_CREATED)


class OrderView(generics.ListAPIView):
    serializer_class = OrderSerializer
    permission_classes = (IsAuthenticated,)

    def get_queryset(self):
        method = self.request.query_params.get("method")
        user = self.request.user

        if method == "for_complete":
            return Order.objects.filter(user=self.request.user, status=1)
        else:
            return Order.objects.filter(user=user)

    def post(self, request, *args, **kwargs):
        addressID = request.data.get('address')
        payment = request.data.get('payment')
        statusID = request.data.get('status')
        orderID = request.data.get('id')
        total = request.data.get('total')
        statusName = OrderStatus.objects.get(pk=statusID)

        if payment:
            order = Order.objects.get(pk=orderID, user=self.request.user)
            if payment == "Cash" or payment == "Card to courier":
                order.comment = f"Payment method: {payment}"
            else:
                paymentID = Payments.objects.get(number=payment, owner=self.request.user)
                order.payment = paymentID
            order.status = statusName
            order.save()
            return Response(status=status.HTTP_205_RESET_CONTENT)
        else:
            address = Address.objects.get(pk=addressID, owner=self.request.user)
            exists = Order.objects.filter(user=self.request.user, status=statusName).exists()
            if exists:
                return Response(status=status.HTTP_303_SEE_OTHER)
            order = Order(user=self.request.user, address=address, total=total, status=statusName)
            order.save()
            return Response(status=status.HTTP_201_CREATED)

    def perform_create(self, serializer):
        if serializer.is_valid():
            serializer.save(owner=self.request.user)
        else:
            return Response(status=status.HTTP_400_BAD_REQUEST)


class ItemListCreate(generics.ListCreateAPIView):
    serializer_class = ItemSerializer
    permission_classes = [IsAuthenticated,]

    def post(self, request, *args, **kwargs):
        method = self.request.data.get('method')
        cartItems = self.request.data.get('items')

        if len(cartItems) > 0:
            if method == "for_total":
                length = len(cartItems)
                total = 0

                for i in cartItems:
                    item = Item.objects.get(pk=i.get("item"))
                    total += item.price * i.get("quantity") + 0.99
                    total = float('{:.2f}'.format(total))
                    if i == cartItems[length - 1]:
                        return Response(status=status.HTTP_200_OK, data={"total": total})
            else:
                return Response(status=status.HTTP_400_BAD_REQUEST)
        else:
            return Response(status=status.HTTP_204_NO_CONTENT)

    def get_queryset(self):
        itemID = self.request.query_params.get("id")

        if itemID:
            return Item.objects.filter(pk=itemID)
        else:
            return Item.objects.all()

    def perform_create(self, serializer):
        if serializer.is_valid():
            serializer.save()
        else:
            print(serializer.errors)


class CheckInCart(generics.ListCreateAPIView):
    serializer_class = CartItemSerializer
    permission_classes = [IsAuthenticated,]

    def list(self, request, *args, **kwargs):
        itemID = self.request.query_params.get("item")
        extra_status = self.request.query_params.get("extra")
        user = self.request.user

        if extra_status == "true":
            return Response(status=status.HTTP_200_OK)

        if itemID:
            isInCart = CartItem.objects.filter(item=itemID, owner=user).exists()

            if isInCart:
                return Response(status=status.HTTP_200_OK)
            else:
                return Response(status=status.HTTP_204_NO_CONTENT)
        return Response(status=status.HTTP_400_BAD_REQUEST)


class DeleteCartItem(generics.DestroyAPIView):
    """
    Deletes a cart item or clears all cart items for the authenticated user.

    Parameters:
    request (Request): The HTTP request object containing query parameters.
    *args: Additional positional arguments.
    **kwargs: Additional keyword arguments.

    Query Parameters:
    - id (str): The ID of the cart item to delete.
    - method (str): The method to use. If "clear", all cart items for the user will be deleted.

    Returns:
    Response:
    - HTTP 202 Accepted if the item(s) were successfully deleted.
    - HTTP 404 Not Found if the specified cart item does not exist.
    - HTTP 400 Bad Request if required query parameters are missing or invalid.
    """
    serializer_class = CartItemSerializer
    permission_classes = [IsAuthenticated,]

    def destroy(self, request, *args, **kwargs):
        itemID = self.request.query_params.get("id")
        method = self.request.query_params.get("method")
        user = self.request.user

        if method == "clear":
            items = CartItem.objects.filter(owner=user)
            for item in items:
                item.delete()
            return Response(status=status.HTTP_202_ACCEPTED)

        if itemID:
            cart_item = CartItem.objects.get(pk=itemID, owner=user)
            if cart_item:
                cart_item.delete()
                return Response(status=status.HTTP_202_ACCEPTED)
            else:
                return Response(status=status.HTTP_404_NOT_FOUND)
        return Response(status=status.HTTP_400_BAD_REQUEST)


class CartItemListCreate(generics.ListCreateAPIView):
    serializer_class = CartItemSerializer
    permission_classes = [IsAuthenticated,]

    def get_queryset(self):
        user = self.request.user
        return CartItem.objects.filter(owner=user)

    def perform_create(self, serializer):
        method = self.request.query_params.get("method")

        if method == "addQuant":
            if serializer.is_valid():
                user = self.request.user
                item = CartItem.objects.get(owner=user, item_id=self.request.data["item"])
                item.quantity = self.request.data["quantity"]
                item.save()
            else:
                print(serializer.errors)
        else:
            if serializer.is_valid():
                serializer.save(owner=self.request.user)
            else:
                print(serializer.errors)


class CouponListCreate(generics.ListCreateAPIView):
    """
    View to list all coupons and create a new coupon.

    - GET: Returns a list of all coupons.
    - POST: Creates a new coupon.

    Permissions:
    - Only authenticated users can access this view.
    """
    serializer_class = CouponSerializer
    permission_classes = [IsAuthenticated,]

    def get_queryset(self):
        method = self.request.query_params.get('method')

        if method == "all":
            return Coupon.objects.all()

        return {"error": "Bad Request"}


    def post(self, request, *args, **kwargs):
        method = self.request.data.get("method")

        if method == "create":
            date = self.request.data.get('date_to')
            title = self.request.data.get('title')
            value = self.request.data.get('value')
            category = self.request.data.get('category')
            enabled_activations = self.request.data.get('ea')
            valid_to = datetime.strptime(date, "%Y-%m-%d").date()
            cat_obj = CouponCategory.objects.get(pk=category)

            check = Coupon.objects.filter(title=title).exists()
            if check:
                return Response(status=status.HTTP_409_CONFLICT)

            if title and value and cat_obj and enabled_activations and valid_to and check == False:
                new_coupon = Coupon(title=title, category=cat_obj, enabled_times=enabled_activations, value=value,
                                    valid_to=valid_to)
                new_coupon.save()
                return Response(status=status.HTTP_201_CREATED)

        if method == "apply":
            title = self.request.data.get('title')
            now = timezone.now().date()
            check = Coupon.objects.filter(title=title,valid_to__gt=now, is_valid=True, enabled_times__gt=0).exists()
            if check:
                coupon = Coupon.objects.get(title=title, valid_to__gt=now, is_valid=True, enabled_times__gt=0)
                coupon.enabled_times -= 1
                coupon.times_activated += 1
                coupon.save()
                return Response(status=status.HTTP_200_OK, data={"value": coupon.value})
            else:
                return Response(status=status.HTTP_204_NO_CONTENT)
        else:
            return Response(status=status.HTTP_400_BAD_REQUEST)


class CancelOrder(generics.ListAPIView):
    serializer_class = OrderSerializer
    permission_classes = [IsAuthenticated,]

    def post(self, request, *args, **kwargs):
        order_id = self.request.data.get('id')
        order_status = self.request.data.get('status')

        order = Order.objects.get(pk=order_id, user=self.request.user)
        status = OrderStatus.objects.get(pk=order_status)

        if order.status == status:
            new_status = OrderStatus.objects.get(pk=5)
            order.status = new_status
            order.save()
            return Response(status=200)
        else:
            return Response(status=400, data={"error": "No order with status " + status})


class RestaurantListCreate(generics.ListCreateAPIView):
    serializer_class = RestaurantSerializer
    permission_classes = [IsAuthenticated,]

    def get_queryset(self):
        return Restaurant.objects.all()

    def perform_create(self, serializer):
        if serializer.is_valid():
            serializer.save()
        else:
            print(serializer.errors)


class SearchItems(generics.ListAPIView):
    serializer_class = ItemSerializer
    permission_classes = [IsAuthenticated,]

    def get_queryset(self):
        query = self.request.query_params.get('q')
        return Item.objects.filter(name__icontains=query)


class CategoriesList(generics.ListCreateAPIView):
    serializer_class = CategorySerializer
    permission_classes = [IsAuthenticated,]

    def get_queryset(self):
        return Category.objects.all()


class RestaurantCategories(generics.ListCreateAPIView):
    serializer_class = RestaurantCatSerializer
    permission_classes = [IsAuthenticated,]

    def get_queryset(self):
        return RestaurantCat.objects.all()


class ItemDelete(generics.DestroyAPIView):
    serializer_class = ItemSerializer
    permission_classes = [IsAuthenticated,]

    def get_queryset(self):
        return Item.objects.all()
    

class CreateUserView(generics.CreateAPIView):
    queryset = User.objects.all()
    serializer_class = UserSerializer
    permission_classes = [AllowAny,]


class SearchView(generics.ListAPIView):
    serializer_class = ItemSerializer
    permission_classes = [IsAuthenticated,]

    def get(self, request, *args, **kwargs):
        search = self.request.query_params.get('search', None)
        also = self.request.query_params.get('also', None)
        search_mode = self.request.query_params.get('search_mode')

        if search is not None:
            if also is not None:
                if search_mode == "food":
                    search1 = Item.objects.filter(Q(title__icontains=search) & Q(title__icontains=also) | Q(title__icontains=search) | Q(title__icontains=also))
                    serializer = ItemSerializer(search1, many=True)
                    return Response(serializer.data)
                else:
                    search1 = Restaurant.objects.filter(Q(name__icontains=search) & Q(name__icontains=also) | Q(name__icontains=search) | Q(name__icontains=also))
                    serializer = RestaurantSerializer(search1, many=True)
                    return Response(serializer.data)
            else:
                if search_mode == "food":
                    items = Item.objects.filter(Q(title__icontains=search))
                    serializer = ItemSerializer(items, many=True)
                    return Response(serializer.data)
                elif search_mode == "rests":
                    rests = Restaurant.objects.filter(Q(name__icontains=search))
                    serializer = RestaurantSerializer(rests, many=True)
                    return Response(serializer.data)
        else:
            return Response(status=status.HTTP_400_BAD_REQUEST)


from .models import Category, Item  # Ensure Category and Item are imported

class FilterView(generics.ListAPIView):
    serializer_class = ItemSerializer
    permission_classes = [IsAuthenticated,]

    def post(self, request, *args, **kwargs):
        data = self.request.data
        filters = data.get('filters')

        if filters:
            category_name = filters.get('category')
            min_cost, max_cost = filters.get('cost', [None, None])
            min_rating = filters.get('rating')

            query = Q()
            if category_name:
                category = Category.objects.get(name=category_name)
                query &= Q(category=category)
            if min_cost is not None and max_cost is not None:
                query &= Q(price__range=(min_cost, max_cost))
            if min_rating is not None:
                query &= Q(rating__gte=min_rating)

            items = Item.objects.filter(query)
            if len(items) > 0:
                serializer = ItemSerializer(items, many=True)
                return Response(status=status.HTTP_200_OK, data=serializer.data)
            else:
                return Response(status=status.HTTP_200_OK, data=[{"not_found": "no items"}])
        else:
            return Response(status=status.HTTP_400_BAD_REQUEST)