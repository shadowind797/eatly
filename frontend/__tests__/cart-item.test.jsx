import React from 'react';
import {render, fireEvent, waitFor, act} from '@testing-library/react';
import api from '../src/api.js';
import {MemoryRouter as Router} from "react-router-dom";
import DishInCart from '../src/components/Cart/CartItem.jsx';

jest.mock('../src/api');

describe("DishInCart component", () => {
    const mockDishData = []
    const item = {
        "id": 1,
        "title": "Chicken Hell",
        "description": "-",
        "price": 12.0,
        "photo": "http://127.0.0.1:8000/media/img/food/ChikenHell.png",
        "category": 2,
        "rating": 4.8
    }

    beforeEach(() => {
        api.get.mockImplementation((url) => {
            if (url.includes('api/items/category/')) {
                return Promise.resolve({data: mockDishData});
            }
            if (url.includes('api/items/')) {
                return Promise.resolve({status: 200, data: item});
            }
            return Promise.reject(new Error('Not found'));
        });
        api.post.mockImplementation((url) => {
            if (url.includes("api/items/cart/add")) {
                return Promise.resolve({status: 201, data: {}});
            }
            return Promise.reject(new Error('Not found'));
        });
    });

    it('renders without crashing', async () => {
        await act(async () => {
            render(<Router><DishInCart cartItem={{id: 1, item: 1}} onChange={() => {
            }}/></Router>);
        });
    });

    it('matches the snapshot', async () => {
        let asFragment;
        await act(async () => {
            const rendered = render(<Router><DishInCart cartItem={{id: 1, item: 1}} onChange={() => {
            }}/></Router>);
            asFragment = rendered.asFragment;
        });

        await waitFor(() => {
            expect(asFragment()).toMatchSnapshot();
        });
        expect(api.get).toHaveBeenCalledWith("api/items/", {"params": {"id": 1}});
    });

    it('calls the onChange function when the DELETE/PLUS/MINUS button is clicked', async () => {
        const mockChange = jest.fn();
        let getByTestId;
        await act(async () => {
            const rendered = render(<Router><DishInCart cartItem={{id: 1, item: 1}} onChange={mockChange}/></Router>);
            getByTestId = rendered.getByTestId;
        });

        await waitFor(() => {
            fireEvent.click(getByTestId('cross'));
            fireEvent.click(getByTestId('plus'));
            fireEvent.click(getByTestId('minus'));
        });

        expect(mockChange).toHaveBeenCalledTimes(3);
    });
});