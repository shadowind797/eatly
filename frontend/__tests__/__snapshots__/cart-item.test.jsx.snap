// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DishInCart component calls the onChange function when the PLUS/MINUS button is clicked 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`DishInCart component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="item"
  >
    <button
      class="cross"
      data-testid="cross"
    >
      <img
        alt=""
        src="test-file-stub"
      />
    </button>
    <div
      class="main"
    >
      <img
        alt=""
        src="undefined/media/http://127.0.0.1:8000/media/img/food/ChikenHell.png"
      />
      <div
        class="info"
      >
        <h4>
          Chicken Hell
        </h4>
        <h5>
          $12.99
        </h5>
      </div>
    </div>
    <div
      class="quant"
    >
      <button
        data-testid="minus"
      >
        <img
          alt=""
          src="test-file-stub"
        />
      </button>
      <div>
        <p>
          3
        </p>
      </div>
      <button
        data-testid="plus"
      >
        <img
          alt=""
          src="test-file-stub"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
